File: 13-roman-to-integer/roman-to-integer.py
class Solution:
    def romanToInt(self, s: str) -> int:
        romanToInt_map = {
            "I" : 1,
            "V" : 5,
            "X" : 10,
            "L" : 50,
            "C" : 100,
            "D" : 500,
            "M" : 1000
        }
        
        total_sum = 0
        last_roman_value = 0
        
        for c in reversed(s):
            summand = romanToInt_map[c]
            if romanToInt_map[c] < last_roman_value:
                summand = -summand

            total_sum += summand
            last_roman_value = romanToInt_map[c]

        return total_sum
        

File: 13-roman-to-integer/roman-to-integer.cpp
class Solution {
public:
    int romanToInt(string s) {
        int number = 0;
        int i = 0;
        while (i < s.length()) {
            if (mapping.find(s.substr(i, 2)) != mapping.end()) {
                number += mapping[s.substr(i, 2)];
                i=i+2;
            } else {
                number += mapping[s.substr(i, 1)];
                i++;
            }
        }
        return number;
    }



    std::map<std::string, int> mapping = {
        {"I", 1},
        {"V", 5},
        {"X", 10},
        {"L", 50},
        {"C", 100},
        {"D", 500},
        {"M", 1000},
        {"IV", 4},
        {"IX", 9},
        {"XL", 40},
        {"XC", 90},
        {"CD", 400},
        {"CM", 900}
    };



};